How many late days are you using for this assignment?
0

In [4]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import StandardScaler

# Part 1: Load the dataset

In [5]:
# Load the given datasets
kidney_cat = pd.read_csv("chronic_kidney_disease_categorical.csv")
kidney_num = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print(kidney_cat.head())
print(kidney_num.head())

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     76411

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The dataset from the UCI Machine Learning Repository contains medical data used to predict whether a patient has chronic kidney disease. Each row in the dataset represents a patient, and each column represents a feature related to their medical condition. The dataset contains a mix of numerical and categorical variables describing symptoms, lab test results, and demographic information. Features: These are the independent variables used to predict the outcome. Examples include blood pressure, age, and hemoglobin levels.
Label: The dependent variable that we aim to predict – in this case, "class", which indicates whether a patient has chronic kidney disease (ckd) or not (notckd).

Examples of Categorical Features (Non-numeric, representing distinct groups or categories):
RBC (Red Blood Cells) 
Hypertension 
Diabetes Mellitus 
Examples of Numerical Features (Continuous or discrete numeric values):
Age
Blood Pressure (bp)
Serum Creatinine (sc)

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [6]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicates = kidney_num.duplicated().sum()
print(f"Total duplicate rows in numerical dataset: {num_duplicates}")

# Drop these duplicate rows
kidney_num = kidney_num.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_duplicates = kidney_cat.duplicated().sum()
print(f"Total duplicate rows in categorical dataset: {cat_duplicates}")

# Print duplicate rows before dropping (if any exist)
if cat_duplicates > 0:
    print("Duplicate rows in categorical dataset:")
    print(kidney_cat[kidney_cat.duplicated()])

# Drop duplicate rows
kidney_cat = kidney_cat.drop_duplicates()


Total duplicate rows in numerical dataset: 30
Total duplicate rows in categorical dataset: 30
Duplicate rows in categorical dataset:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
231     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
401     628271  2.0  1.0  abnormal    normal  notpresent  notpresent  yes   
402     158929  1.0  0.0       NaN       NaN  notpresent  notpresent   no   
403     240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
404     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
405     337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
406     118191  0.0  0.0    normal    normal  notpresent  notpresent   no   
407     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
408     302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
409     938027  0.0  0.0    normal    normal  notpresent  notpresent   no   
410     572140  0.0 

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: I don't think the unique ids are present in both datasets. If some IDs are missing in one dataset, here are possible reasons:
Incomplete Data Collection, Data Entry Errors, Different Data Sources, Data Cleaning/Processing Issues

In [7]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
if 'unique_id' in kidney_cat.columns and 'unique_id' in kidney_num.columns:
    # Merge both datasets based on 'id'
    kidney_combined = pd.merge(kidney_num, kidney_cat, on="unique_id", how="inner")  # Only keeps matching IDs
    
    # Print the combined dataset
    print(kidney_combined.head())
    
    # Print the shape of the merged dataset to check dimensions
    print(f"Shape of combined dataset: {kidney_combined.shape}")
else:
    print("No 'id' column found in one or both datasets.")

#Print the combined dataset


   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   

       pc         pcc          ba  htn   dm cad appet   pe ane  Target  
0  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
1  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
2  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
3  normal  notpresent  notpresent  yes  yes  no  poor  yes  no     ckd  
4  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  

[5 rows x 25 columns]
Shape of combined dataset: (390, 25)


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [8]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = kidney_combined.isnull().any(axis=1).sum()
total_rows = kidney_combined.shape[0]
missing_percentage = (missing_rows / total_rows) * 100
# Print %
print(f"Percentage of rows with at least one missing value: {missing_percentage:.2f}%")
# Drop these rows from the dataset
kidney_cleaned = kidney_combined.dropna()
# Print the Dataset
print(f"New shape of dataset: {kidney_cleaned.shape}")

Percentage of rows with at least one missing value: 60.77%
New shape of dataset: (153, 25)


## Part 3.4: Sort the dataset according to the Labels

In [9]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
kidney_sorted = kidney_cleaned.sort_values(by="Target").reset_index(drop=True)
    
    # Print the sorted dataset
print(kidney_sorted)
    
    # Print the shape after sorting
print(f"Dataset shape after sorting: {kidney_sorted.shape}")
# Print the dataset

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [10]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
kidney_encoded = pd.get_dummies(kidney_sorted, columns=categorical_columns, drop_first=False)

# Print the dataset
print(kidney_encoded.head())


   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   

   cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0    True    False        True       False   True   False   False     True   
1   False     True       False        True   True   False    True    False   
2    True    False       False        True   True   False   False     True   
3   False     True       False        True  False    True   False     True   
4    True    False        True       False   True   False    True    False   

   Target_ckd  Target_notckd  
0        True          False  
1     

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For binary categorical variables, both one-hot encoding and binary encoding are equivalent, as both provide the same information to the model with one binary column. However, for multi-class categorical variables (e.g., "type of condiment" with 5 values), one-hot encoding creates a separate column for each category, which could lead to a higher-dimensional dataset, but without introducing any ordinal relationships between categories. Binary encoding, on the other hand, reduces dimensionality by representing categories with fewer binary columns, but it introduces an implicit ordinal relationship, which may mislead some models (like linear models) into assuming a distance between categories. For most models, one-hot encoding is preferred for multi-class variables, as it avoids these issues and is simpler to interpret, especially when there is no inherent order in the categories.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [11]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    mean = kidney_sorted[col].mean()
    std_dev = kidney_sorted[col].std()
    
    # Define the outlier threshold (3 times the standard deviation from the mean)
    lower_bound = mean - 3 * std_dev
    upper_bound = mean + 3 * std_dev
    
    # Filter the data to remove outliers
    kidney_sorted = kidney_sorted[(kidney_sorted[col] >= lower_bound) & (kidney_sorted[col] <= upper_bound)]

# Print the dataset
print(kidney_sorted)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [12]:
# Normalize the all Numerical Attributes in the dataset.
scaler = StandardScaler()
kidney_sorted[numerical_columns] = scaler.fit_transform(kidney_sorted[numerical_columns])

# Print the dataset after normalization
print(kidney_sorted)
# Print the dataset

     unique_id       age        bp       bgr        bu        sc       sod  \
0       474407 -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258   
1       137148  1.567253  2.776102  3.853379  1.837868  2.836183 -0.079424   
3       343710  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702   
4       532520  0.747101  1.753327 -0.362349  0.449451  0.611338 -0.783646   
6       546225 -0.136138 -1.314996  0.924557  1.912917  1.285533  0.096632   
..         ...       ...       ...       ...       ...       ...       ...   
148     647552  0.810190 -0.292221 -0.029528 -0.451144 -0.467375 -0.607591   
149     125587 -0.703936 -1.314996 -0.273597  0.224302 -0.197696  0.096632   
150     614376  0.557836  0.730553 -0.051716 -0.901441 -0.130277 -0.959702   
151     853426 -0.451581 -0.292221  0.081412 -0.113421 -0.534794 -0.079424   
152     572140  0.431659  0.730553 -0.384537 -0.488668 -0.332535  0.272688   

          pot      hemo       pcv  ...        pc         pcc   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Some binary columns like ba, htn, dm, etc., might be better represented through dummy variables or could be included after proper transformation. However, if any of these columns do not provide significant variation or meaningful information, they could be excluded. Justification: These columns might already be captured in a different form, such as through dummy encoding.

In [13]:
#Remove that column
columns_to_drop = ['ba']

# Remove the columns from the dataset
kidney_model_ready = kidney_sorted.drop(columns=columns_to_drop)

# Print the cleaned dataset
print(kidney_model_ready)
# Print the dataset

     unique_id       age        bp       bgr        bu        sc       sod  \
0       474407 -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258   
1       137148  1.567253  2.776102  3.853379  1.837868  2.836183 -0.079424   
3       343710  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702   
4       532520  0.747101  1.753327 -0.362349  0.449451  0.611338 -0.783646   
6       546225 -0.136138 -1.314996  0.924557  1.912917  1.285533  0.096632   
..         ...       ...       ...       ...       ...       ...       ...   
148     647552  0.810190 -0.292221 -0.029528 -0.451144 -0.467375 -0.607591   
149     125587 -0.703936 -1.314996 -0.273597  0.224302 -0.197696  0.096632   
150     614376  0.557836  0.730553 -0.051716 -0.901441 -0.130277 -0.959702   
151     853426 -0.451581 -0.292221  0.081412 -0.113421 -0.534794 -0.079424   
152     572140  0.431659  0.730553 -0.384537 -0.488668 -0.332535  0.272688   

          pot      hemo       pcv  ...       rbc        pc     

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [14]:
# Export the dataframe to a new csv file
kidney_model_ready.to_csv("cleaned_kidney_data.csv", index=False)
kidney_model_ready.to_json("cleaned_kidney_data.json", orient="records", lines=True)

# Print confirmation message
print("Data has been exported to 'cleaned_kidney_data.csv' and 'cleaned_kidney_data.json'.")
# Export the dataframe to a new json file

Data has been exported to 'cleaned_kidney_data.csv' and 'cleaned_kidney_data.json'.


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [15]:
kidney_model_ready.head(15)

# Print the top 15 rows to the terminal

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,rbc,pc,pcc,htn,dm,cad,appet,pe,ane,Target
0,474407,-1.713353,1.753327,-0.317973,-0.038371,0.206821,-2.720258,-1.303189,-2.652160,-3.050180,...,normal,abnormal,present,no,no,no,good,no,yes,ckd
1,137148,1.567253,2.776102,3.853379,1.837868,2.836183,-0.079424,-2.247736,-2.260568,-2.037442,...,abnormal,abnormal,present,yes,yes,yes,poor,no,no,ckd
3,343710,0.810190,0.730553,1.146438,4.014305,1.690051,-0.959702,1.373024,-2.913221,-2.905503,...,abnormal,abnormal,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,0.747101,1.753327,-0.362349,0.449451,0.611338,-0.783646,1.373024,-1.433874,-1.603411,...,abnormal,normal,notpresent,no,no,no,good,no,no,ckd
6,546225,-0.136138,-1.314996,0.924557,1.912917,1.285533,0.096632,-0.516068,-1.999507,-2.326796,...,normal,normal,notpresent,yes,yes,no,good,no,no,ckd
8,514721,0.494747,1.753327,0.170164,2.475789,3.577798,-1.663925,0.743327,-2.304078,-2.182119,...,abnormal,abnormal,notpresent,yes,no,no,good,no,no,ckd
10,621332,0.684013,0.730553,4.030883,-0.225995,-0.062857,-3.248425,-1.303189,-1.738445,-1.314058,...,abnormal,normal,notpresent,no,yes,no,poor,no,no,ckd
13,481293,0.747101,-1.314996,3.698062,-0.188470,0.206821,-1.839980,-2.090311,-2.826200,-2.760826,...,normal,abnormal,present,yes,no,no,poor,no,yes,ckd
17,615697,-2.659681,-1.314996,-0.606418,0.974798,-0.265116,-0.959702,0.900751,-1.955996,-2.037442,...,abnormal,abnormal,notpresent,no,no,no,poor,no,no,ckd
19,843362,1.441076,-1.314996,-0.073905,3.151235,2.633924,-0.783646,0.900751,-1.303343,-1.314058,...,normal,normal,notpresent,yes,yes,no,poor,yes,no,ckd


Paste here: 
|   | unique_id | age      | bp        | bgr        | bu        | sc        | sod       | pot        | hemo      | pcv        | ...       | rbc     | pc        | pcc        | htn | dm  | cad | appet | pe   | ane   | Target |
|---|-----------|----------|-----------|------------|-----------|-----------|-----------|------------|-----------|------------|-----------|---------|-----------|------------|-----|-----|-----|-------|------|-------|--------|
| 0 | 474407    | -1.713353| 1.753327  | -0.317973  | -0.038371 | 0.206821  | -2.720258 | -1.303189  | -2.652160 | -3.050180  | ...       | normal  | abnormal  | present    | no  | no  | no  | good  | no   | yes   | ckd    |
| 1 | 137148    | 1.567253 | 2.776102  | 3.853379   | 1.837868  | 2.836183  | -0.079424 | -2.247736  | -2.260568 | -2.037442  | ...       | abnormal| abnormal  | present    | yes | yes | yes | poor  | no   | no    | ckd    |
| 3 | 343710    | 0.810190 | 0.730553  | 1.146438   | 4.014305  | 1.690051  | -0.959702 | 1.373024   | -2.913221 | -2.905503  | ...       | abnormal| abnormal  | notpresent | yes | yes | yes | poor  | yes  | yes   | ckd    |
| 4 | 532520    | 0.747101 | 1.753327  | -0.362349  | 0.449451  | 0.611338  | -0.783646 | 1.373024   | -1.433874 | -1.603411  | ...       | abnormal| normal    | notpresent | no  | no  | no  | good  | no   | no    | ckd    |
| 6 | 546225    | -0.136138| -1.314996 | 0.924557   | 1.912917  | 1.285533  | 0.096632  | -0.516068  | -1.999507 | -2.326796  | ...       | normal  | normal    | notpresent | yes | yes | no  | good  | no   | no    | ckd    |
| 8 | 514721    | 0.494747 | 1.753327  | 0.170164   | 2.475789  | 3.577798  | -1.663925 | 0.743327   | -2.304078 | -2.182119  | ...       | abnormal| abnormal  | notpresent | yes | no  | no  | good  | no   | no    | ckd    |
| 10| 621332    | 0.684013 | 0.730553  | 4.030883   | -0.225995 | -0.062857 | -3.248425 | -1.303189  | -1.738445 | -1.314058  | ...       | abnormal| normal    | notpresent | no  | yes | no  | poor  | no   | no    | ckd    |
| 13| 481293    | 0.747101 | -1.314996 | 3.698062   | -0.188470 | 0.206821  | -1.839980 | -2.090311  | -2.826200 | -2.760826  | ...       | normal  | abnormal  | present    | yes | no  | no  | poor  | no   | yes   | ckd    |
| 17| 615697    | -2.659681| -1.314996 | -0.606418  | 0.974798  | -0.265116 | -0.959702 | 0.900751   | -1.955996 | -2.037442  | ...       | abnormal| abnormal  | notpresent | no  | no  | no  | poor  | no   | no    | ckd    |
| 19| 843362    | 1.441076 | -1.314996 | -0.073905  | 3.151235  | 2.633924  | -0.783646 | 0.900751   | -1.303343 | -1.314058  | ...       | normal  | normal    | notpresent | yes | yes | no  | poor  | yes  | no    | ckd    |
| 20| 995177    | 1.314898 | -0.292221 | 2.056147   | 2.063016  | 3.308119  | -3.600537 | -0.673492  | -2.173547 | -2.326796  | ...       | normal  | abnormal  | present    | yes | yes | yes | good  | yes  | yes   | ckd    |
| 21| 828592    | 1.441076 | -0.292221 | 2.167088   | 1.537669  | 1.487792  | -1.311813 | 0.113630   | -1.738445 | -1.603411  | ...       | normal  | abnormal  | present    | yes | yes | yes | good  | no   | no    | ckd    |
| 22| 923613    | 0.999456 | -1.314996 | 2.610848   | 0.637075  | 1.959729  | -0.607591 | 1.687873   | -2.130037 | -2.182119  | ...       | abnormal| abnormal  | notpresent | yes | yes | no  | poor  | yes  | no    | ckd    |
| 24| 955830    | 0.431659 | 0.730553  | 2.056147   | 1.199946  | 1.690051  | -0.607591 | 0.900751   | -1.520894 | -1.458735  | ...       | normal  | abnormal  | present    | yes | yes | no  | good  | yes  | no    | ckd    |
| 25| 397388    | 0.936367 | 2.776102  | 3.520558   | -0.225995 | 1.218114  | 0.448743  | -1.303189  | -0.607180 | -0.590673  | ...       | normal  | normal    | notpresent | yes | no  | yes | go


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [ ]:
# Code to rename all the columns in the dataset
# Create a dictionary for the column name mapping
column_mapping = {
    'age': 'Age (years)',
    'bp': 'Blood Pressure (mm/Hg)',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random (mg/dl)',
    'bu': 'Blood Urea (mg/dl)',
    'sc': 'Serum Creatinine (mg/dl)',
    'sod': 'Sodium (mEq/L)',
    'pot': 'Potassium (mEq/L)',
    'hemo': 'Hemoglobin (gms)',
    'pcv': 'Packed Cell Volume',
    'wc': 'White Blood Cell Count (cells/cumm)',
    'rc': 'Red Blood Cell Count (millions/cmm)',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'Target': 'Chronic Kidney Disease (CKD)'  # Renamed 'class' to 'Target'
}


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [17]:
# Converted SQL to Pandas code
result = kidney_model_ready.groupby('Target').size().reset_index(name='count')

*Describe what the above code does here*
This pandas code groups the dataset by the 'Target' column and counts the number of occurrences for each unique value in 'Target'. The resulting output will show the 'Target' values along with a count of how many times each appears in the dataset.

If you run the above pandas code, you will get a dataframe with two columns: 'Target' and 'count'.